File: 2301_match_substring_after_replacement/solution.ts
// LeetCode 2301 - Match Substring After Replacement
// https://leetcode.com/problems/match-substring-after-replacement/

export function matchReplacement(s: any, sub: any, mappings: any): any {
    const allow = new Set();
    for (const m of mappings) allow.add((m[0].charCodeAt(0) << 8) | m[1].charCodeAt(0));
    const n = s.length, mlen = sub.length;
    for (let i = 0; i + mlen <= n; i++) {
        let ok = true;
        for (let j = 0; j < mlen; j++) {
            const a = s[i + j], b = sub[j];
            if (a === b || allow.has((b.charCodeAt(0) << 8) | a.charCodeAt(0))) continue;
            ok = false;
            break;
        }
        if (ok) return true;
    }
    return false;
}

File: 2301_match_substring_after_replacement/Solution.swift
// LeetCode 2301 - Match Substring After Replacement
// https://leetcode.com/problems/match-substring-after-replacement/

class Solution {
    func matchReplacement(_ s: String, _ sub: String, _ mappings: [[Character]]) -> Bool {
        var allow = Set<Int>()
        for m in mappings {
            allow.insert((Int(m[0].asciiValue!) << 8) | Int(m[1].asciiValue!))
        }
        let sArr = Array(s), subArr = Array(sub)
        let n = sArr.count, mlen = subArr.count
        if n < mlen { return false }
        for i in 0...(n - mlen) {
            var ok = true
            for j in 0..<mlen {
                let a = sArr[i + j], b = subArr[j]
                if a == b || allow.contains((Int(b.asciiValue!) << 8) | Int(a.asciiValue!)) { continue }
                ok = false
                break
            }
            if ok { return true }
        }
        return false
    }
}
